content/blog/2022-04-18-terraform-provider.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Machine Learning Workloads with Iterative Provider for Terraform
+date: 2022-04-18
+description: >
+  Today we introduce resource orchestration for your machine learning
+  projects  in conjunction with Terraform.
+descriptionLong: >
+  Now you can bring multi-cloud provisioning to your machine learning projects,
+  optimizing resources; costs, and workflows while aligning your Data
+  Science  and DevOps teams.
+picture: 2022-04-18/terraform-provider-iterative.png
+pictureComment: Machine Learning Workloads with Iterative Provider for Terraform
+author: maria_khalusova
+commentsUrl: https://discuss.dvc.org/t/collaborative-experiments/1002
+tags:
+  - MLOps
+  - Terraform
+  - Git
+  - Resource orchestation
+  - Spot Instance management
+---
+
+The requirements for Machine Learning (ML) infrastructure are becoming
+increasingly complex. Training large models often requires specialized hardware
+(GPUs, TPUs) which involves moving the whole training process onto cloud
+machines, setting up environments and synchronizing data. For teams that want to
+leverage spot instances, the setup becomes even more complex - they need to make
+sure the training progress is not lost during spot instance recovery. This is
+time-consuming, and requires expertise in both DevOps and Machine Learning. On
+top of that, the costs of training in a cloud environment can be high due to not
+only the need for specific hardware, but also individual responsibility to shut
+down instances when training is complete.
+
+To address the specific needs of machine learning teams, we have built the
+Iterative Provider (TPI). TPI is an open-source tool extending the functionality
+of Terraform, the world's most widely used multi-cloud provisioning product. The
+Iterative Provider enables full lifecycle management of computing resources and
+is designed specifically for machine learning pipelines.
+
+## Tailored to Machine Learning Workflows
+
+The Iterative Provider offers a single resource called `iterative_task` which
+you can use to configure:
+
+- Your cloud infrastructure
+- The steps to perform on the cloud resource, i.e. setting up the environment,
+  running the training pipeline, logging metrics, etc.
+- The data to be synced back once the training is complete (e.g. a file with
+  metrics, a model, plots)
+
+Here’s a “hello world” example of a `main.tf` Terraform configuration file using
+the `iterative_task`:
+
+```hcl
+terraform {
+     required_providers {
+         iterative = {source = "iterative/iterative", version = ">= 0.9.9" }
+     }
+ }
+ provider "iterative" {}
+
+ resource "iterative_task" "hello_world" {
+     name      = "ml_project_1"
+     cloud     = "aws"
+     region    = "us-east-2"
+     machine   = "l+k80"
+     directory = "."
+     script = <<-END
+       #!/bin/bash
+       sudo apt update
+       sudo apt install -y python3-pip
+       pip3 install -r requirements.txt
+       python3 src/train.py
+     END
+ }
```

Comment:
Updated as per the [Getting Started](https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/getting-started) guide.

```suggestion
terraform {
  required_providers {
    iterative = {
      source = "iterative/iterative"
    }
  }
}

provider "iterative" {}

resource "iterative_task" "example" {
  cloud   = "aws"
  region  = "us-east-2"
  machine = "l+k80"
  
  script = <<-END
    #!/bin/bash
    sudo apt update
    sudo apt install -y python3-pip
    pip3 install -r requirements.txt
    python3 src/train.py
  END
  
  storage {
    workdir = "."
    output  = "results"
  }
}
```

🔔 @casperdcl may want to “casperify” the code format.